File: compatibility-tests/sdk-test-java/src/test/java/com/floci/test/EventBridgeTest.java
package com.floci.test;

import org.junit.jupiter.api.*;
import software.amazon.awssdk.services.eventbridge.EventBridgeClient;
import software.amazon.awssdk.services.eventbridge.model.*;
import software.amazon.awssdk.services.sqs.SqsClient;
import software.amazon.awssdk.services.sqs.model.*;

import java.util.List;
import java.util.Map;

import static org.assertj.core.api.Assertions.*;

@DisplayName("EventBridge")
@TestMethodOrder(MethodOrderer.OrderAnnotation.class)
class EventBridgeTest {

    private static EventBridgeClient eb;
    private static SqsClient sqs;

    private static String ruleName;
    private static String busName;
    private static String sinkQueueUrl;
    private static String sinkQueueArn;
    private static String transformerQueueUrl;
    private static String transformerQueueArn;

    @BeforeAll
    static void setup() {
        eb = TestFixtures.eventBridgeClient();
        sqs = TestFixtures.sqsClient();
        ruleName = TestFixtures.uniqueName("eb-rule");
        busName = TestFixtures.uniqueName("eb-bus");
    }

    @AfterAll
    static void cleanup() {
        try {
            eb.removeTargets(RemoveTargetsRequest.builder().rule(ruleName).ids("sqs-target", "transformer-target").build());
        } catch (Exception ignored) {}
        try {
            eb.deleteRule(DeleteRuleRequest.builder().name(ruleName).build());
        } catch (Exception ignored) {}
        try {
            eb.deleteEventBus(DeleteEventBusRequest.builder().name(busName).build());
        } catch (Exception ignored) {}
        try {
            sqs.deleteQueue(DeleteQueueRequest.builder().queueUrl(sinkQueueUrl).build());
        } catch (Exception ignored) {}
        try {
            sqs.deleteQueue(DeleteQueueRequest.builder().queueUrl(transformerQueueUrl).build());
        } catch (Exception ignored) {}
        eb.close();
        sqs.close();
    }

    // ──────────────────────────── Event Buses ────────────────────────────

    @Test
    @Order(1)
    void createEventBus() {
        CreateEventBusResponse response = eb.createEventBus(
                CreateEventBusRequest.builder().name(busName).build());
        assertThat(response.eventBusArn()).contains(busName);
    }

    @Test
    @Order(2)
    void describeEventBus() {
        DescribeEventBusResponse response = eb.describeEventBus(
                DescribeEventBusRequest.builder().name(busName).build());
        assertThat(response.name()).isEqualTo(busName);
        assertThat(response.arn()).contains(busName);
    }

    @Test
    @Order(3)
    void listEventBuses() {
        ListEventBusesResponse response = eb.listEventBuses(
                ListEventBusesRequest.builder().build());
        assertThat(response.eventBuses()).extracting(EventBus::name).contains("default", busName);
    }

    // ──────────────────────────── Rules ────────────────────────────

    @Test
    @Order(10)
    void putRule() {
        PutRuleResponse response = eb.putRule(PutRuleRequest.builder()
                .name(ruleName)
                .eventPattern("{\"source\":[\"com.myapp\"]}")
                .state(RuleState.ENABLED)
                .description("Test rule")
                .build());
        assertThat(response.ruleArn()).contains(ruleName);
    }

    @Test
    @Order(11)
    void describeRule() {
        DescribeRuleResponse response = eb.describeRule(
                DescribeRuleRequest.builder().name(ruleName).build());
        assertThat(response.name()).isEqualTo(ruleName);
        assertThat(response.state()).isEqualTo(RuleState.ENABLED);
        assertThat(response.eventPattern()).contains("com.myapp");
    }

    @Test
    @Order(12)
    void listRules() {
        ListRulesResponse response = eb.listRules(ListRulesRequest.builder().build());
        assertThat(response.rules()).extracting(Rule::name).contains(ruleName);
    }

    @Test
    @Order(13)
    void disableAndEnableRule() {
        eb.disableRule(DisableRuleRequest.builder().name(ruleName).build());
        assertThat(eb.describeRule(DescribeRuleRequest.builder().name(ruleName).build()).state())
                .isEqualTo(RuleState.DISABLED);

        eb.enableRule(EnableRuleRequest.builder().name(ruleName).build());
        assertThat(eb.describeRule(DescribeRuleRequest.builder().name(ruleName).build()).state())
                .isEqualTo(RuleState.ENABLED);
    }

    // ──────────────────────────── Targets + PutEvents ────────────────────────────

    @Test
    @Order(20)
    void createSinkQueue() {
        sinkQueueUrl = sqs.createQueue(CreateQueueRequest.builder()
                .queueName(TestFixtures.uniqueName("eb-sink")).build()).queueUrl();
        sinkQueueArn = sqs.getQueueAttributes(GetQueueAttributesRequest.builder()
                .queueUrl(sinkQueueUrl).attributeNamesWithStrings("QueueArn").build())
                .attributesAsStrings().get("QueueArn");
        assertThat(sinkQueueArn).contains("eb-sink");
    }

    @Test
    @Order(21)
    void putSqsTarget() {
        PutTargetsResponse response = eb.putTargets(PutTargetsRequest.builder()
                .rule(ruleName)
                .targets(Target.builder().id("sqs-target").arn(sinkQueueArn).build())
                .build());
        assertThat(response.failedEntryCount()).isZero();
    }

    @Test
    @Order(22)
    void listTargetsByRule() {
        ListTargetsByRuleResponse response = eb.listTargetsByRule(
                ListTargetsByRuleRequest.builder().rule(ruleName).build());
        assertThat(response.targets()).extracting(Target::id).contains("sqs-target");
    }

    @Test
    @Order(23)
    void putEventsDeliveredToSqsTarget() {
        eb.putEvents(PutEventsRequest.builder()
                .entries(PutEventsRequestEntry.builder()
                        .source("com.myapp")
                        .detailType("OrderPlaced")
                        .detail("{\"orderId\":\"123\"}")
                        .build())
                .build());

        ReceiveMessageResponse msg = sqs.receiveMessage(ReceiveMessageRequest.builder()
                .queueUrl(sinkQueueUrl)
                .maxNumberOfMessages(1)
                .waitTimeSeconds(2)
                .build());
        assertThat(msg.messages()).hasSize(1);
        assertThat(msg.messages().get(0).body()).contains("com.myapp").contains("OrderPlaced");
    }

    @Test
    @Order(24)
    void putEventsNoMatchingRuleNotDelivered() {
        eb.putEvents(PutEventsRequest.builder()
                .entries(PutEventsRequestEntry.builder()
                        .source("other.app")
                        .detailType("Ignored")
                        .detail("{}")
                        .build())
                .build());

        ReceiveMessageResponse msg = sqs.receiveMessage(ReceiveMessageRequest.builder()
                .queueUrl(sinkQueueUrl)
                .maxNumberOfMessages(1)
                .build());
        assertThat(msg.messages()).isEmpty();
    }

    @Test
    @Order(25)
    void putEventsWithPrefixPatternDeliveredToSqsTarget() {
        String prefixRuleName = TestFixtures.uniqueName("eb-prefix-rule");
        eb.putRule(PutRuleRequest.builder()
                .name(prefixRuleName)
                .eventPattern("{\"source\":[{\"prefix\":\"com.example\"}]}")
                .state(RuleState.ENABLED)
                .build());

        eb.putTargets(PutTargetsRequest.builder()
                .rule(prefixRuleName)
                .targets(Target.builder().id("prefix-sqs-target").arn(sinkQueueArn).build())
                .build());

        // Matching source
        eb.putEvents(PutEventsRequest.builder()
                .entries(PutEventsRequestEntry.builder()
                        .source("com.example.myapp")
                        .detailType("Test")
                        .detail("{}")
                        .build())
                .build());

        ReceiveMessageResponse msg = sqs.receiveMessage(ReceiveMessageRequest.builder()
                .queueUrl(sinkQueueUrl)
                .maxNumberOfMessages(1)
                .waitTimeSeconds(2)
                .build());
        
        try {
            assertThat(msg.messages()).hasSize(1);
            assertThat(msg.messages().get(0).body()).contains("com.example.myapp");
        } finally {
            // Cleanup for this specific test
            eb.removeTargets(RemoveTargetsRequest.builder().rule(prefixRuleName).ids("prefix-sqs-target").build());
            eb.deleteRule(DeleteRuleRequest.builder().name(prefixRuleName).build());
        }
    }

    // ──────────────────────────── InputTransformer ────────────────────────────

    @Test
    @Order(30)
    void createTransformerQueue() {
        transformerQueueUrl = sqs.createQueue(CreateQueueRequest.builder()
                .queueName(TestFixtures.uniqueName("eb-xform")).build()).queueUrl();
        transformerQueueArn = sqs.getQueueAttributes(GetQueueAttributesRequest.builder()
                .queueUrl(transformerQueueUrl).attributeNamesWithStrings("QueueArn").build())
                .attributesAsStrings().get("QueueArn");
        assertThat(transformerQueueArn).contains("eb-xform");
    }

    @Test
    @Order(31)
    void putInputTransformerTarget() {
        PutTargetsResponse response = eb.putTargets(PutTargetsRequest.builder()
                .rule(ruleName)
                .targets(Target.builder()
                        .id("transformer-target")
                        .arn(transformerQueueArn)
                        .inputTransformer(InputTransformer.builder()
                                .inputPathsMap(Map.of("src", "$.source", "type", "$.detail-type"))
                                .inputTemplate("{\"source\":\"<src>\",\"type\":\"<type>\"}")
                                .build())
                        .build())
                .build());
        assertThat(response.failedEntryCount()).isZero();
    }

    @Test
    @Order(32)
    void inputTransformerTargetStoredCorrectly() {
        ListTargetsByRuleResponse response = eb.listTargetsByRule(
                ListTargetsByRuleRequest.builder().rule(ruleName).build());
        Target xformTarget = response.targets().stream()
                .filter(t -> "transformer-target".equals(t.id()))
                .findFirst().orElseThrow();
        assertThat(xformTarget.inputTransformer()).isNotNull();
        assertThat(xformTarget.inputTransformer().inputPathsMap()).containsKey("src");
        assertThat(xformTarget.inputTransformer().inputTemplate()).contains("<src>");
    }

    @Test
    @Order(33)
    void putEventsInputTransformerTransformsPayload() {
        // Drain any prior messages
        sqs.receiveMessage(ReceiveMessageRequest.builder()
                .queueUrl(transformerQueueUrl).maxNumberOfMessages(10).build());

        eb.putEvents(PutEventsRequest.builder()
                .entries(PutEventsRequestEntry.builder()
                        .source("com.myapp")
                        .detailType("OrderShipped")
                        .detail("{\"orderId\":\"456\"}")
                        .build())
                .build());

        ReceiveMessageResponse msg = sqs.receiveMessage(ReceiveMessageRequest.builder()
                .queueUrl(transformerQueueUrl)
                .maxNumberOfMessages(1)
                .waitTimeSeconds(2)
                .build());
        assertThat(msg.messages()).hasSize(1);
        String body = msg.messages().get(0).body();
        assertThat(body).contains("com.myapp").contains("OrderShipped");
        assertThat(body).doesNotContain("orderId");
    }

    // ──────────────────────────── Tags ────────────────────────────

    @Test
    @Order(40)
    void listTagsForResource() {
        String ruleArn = eb.describeRule(DescribeRuleRequest.builder().name(ruleName).build()).arn();
        ListTagsForResourceResponse response = eb.listTagsForResource(
                ListTagsForResourceRequest.builder().resourceARN(ruleArn).build());
        assertThat(response.tags()).isNotNull();
    }

    // ──────────────────────────── Cleanup ────────────────────────────

    @Test
    @Order(90)
    void removeTargets() {
        RemoveTargetsResponse response = eb.removeTargets(RemoveTargetsRequest.builder()
                .rule(ruleName)
                .ids("sqs-target", "transformer-target")
                .build());
        assertThat(response.failedEntryCount()).isZero();
        assertThat(eb.listTargetsByRule(ListTargetsByRuleRequest.builder().rule(ruleName).build())
                .targets()).isEmpty();
    }

    @Test
    @Order(91)
    void deleteRule() {
        eb.deleteRule(DeleteRuleRequest.builder().name(ruleName).build());
        assertThatThrownBy(() ->
                eb.describeRule(DescribeRuleRequest.builder().name(ruleName).build()))
                .isInstanceOf(software.amazon.awssdk.services.eventbridge.model.ResourceNotFoundException.class);
    }

    @Test
    @Order(92)
    void deleteEventBus() {
        eb.deleteEventBus(DeleteEventBusRequest.builder().name(busName).build());
        ListEventBusesResponse response = eb.listEventBuses(ListEventBusesRequest.builder().build());
        assertThat(response.eventBuses()).extracting(EventBus::name).doesNotContain(busName);
    }
}

File: compatibility-tests/sdk-test-node/tests/eventbridge.test.ts
/**
 * EventBridge integration tests.
 */

import { describe, it, expect, beforeAll, afterAll } from 'vitest';
import {
  EventBridgeClient,
  CreateEventBusCommand,
  DeleteEventBusCommand,
  DescribeEventBusCommand,
  ListEventBusesCommand,
  PutRuleCommand,
  DeleteRuleCommand,
  DescribeRuleCommand,
  ListRulesCommand,
  EnableRuleCommand,
  DisableRuleCommand,
  PutTargetsCommand,
  RemoveTargetsCommand,
  ListTargetsByRuleCommand,
  PutEventsCommand,
  ListTagsForResourceCommand,
} from '@aws-sdk/client-eventbridge';
import {
  SQSClient,
  CreateQueueCommand,
  DeleteQueueCommand,
  GetQueueAttributesCommand,
  ReceiveMessageCommand,
} from '@aws-sdk/client-sqs';
import { makeClient, uniqueName } from './setup';

describe('EventBridge', () => {
  let eb: EventBridgeClient;
  let sqs: SQSClient;
  let ruleName: string;
  let busName: string;
  let ruleArn: string;
  let sinkQueueUrl: string;
  let sinkQueueArn: string;
  let transformerQueueUrl: string;
  let transformerQueueArn: string;

  beforeAll(() => {
    eb = makeClient(EventBridgeClient);
    sqs = makeClient(SQSClient);
    ruleName = uniqueName('eb-rule');
    busName = uniqueName('eb-bus');
  });

  afterAll(async () => {
    try {
      await eb.send(new RemoveTargetsCommand({ Rule: ruleName, Ids: ['sqs-target', 'transformer-target'] }));
    } catch { /* ignore */ }
    try {
      await eb.send(new DeleteRuleCommand({ Name: ruleName }));
    } catch { /* ignore */ }
    try {
      await eb.send(new DeleteEventBusCommand({ Name: busName }));
    } catch { /* ignore */ }
    try {
      if (sinkQueueUrl) await sqs.send(new DeleteQueueCommand({ QueueUrl: sinkQueueUrl }));
    } catch { /* ignore */ }
    try {
      if (transformerQueueUrl) await sqs.send(new DeleteQueueCommand({ QueueUrl: transformerQueueUrl }));
    } catch { /* ignore */ }
  });

  // ──────────────────────────── Event Buses ────────────────────────────

  it('should create an event bus', async () => {
    const response = await eb.send(new CreateEventBusCommand({ Name: busName }));
    expect(response.EventBusArn).toContain(busName);
  });

  it('should describe an event bus', async () => {
    const response = await eb.send(new DescribeEventBusCommand({ Name: busName }));
    expect(response.Name).toBe(busName);
    expect(response.Arn).toContain(busName);
  });

  it('should list event buses including default', async () => {
    const response = await eb.send(new ListEventBusesCommand({}));
    const names = response.EventBuses!.map(b => b.Name);
    expect(names).toContain('default');
    expect(names).toContain(busName);
  });

  // ──────────────────────────── Rules ────────────────────────────

  it('should create a rule', async () => {
    const response = await eb.send(new PutRuleCommand({
      Name: ruleName,
      EventPattern: JSON.stringify({ source: ['com.myapp'] }),
      State: 'ENABLED',
      Description: 'Test rule',
    }));
    ruleArn = response.RuleArn!;
    expect(ruleArn).toContain(ruleName);
  });

  it('should describe the rule', async () => {
    const response = await eb.send(new DescribeRuleCommand({ Name: ruleName }));
    expect(response.Name).toBe(ruleName);
    expect(response.State).toBe('ENABLED');
    expect(response.EventPattern).toContain('com.myapp');
  });

  it('should list rules', async () => {
    const response = await eb.send(new ListRulesCommand({}));
    const names = response.Rules!.map(r => r.Name);
    expect(names).toContain(ruleName);
  });

  it('should disable and re-enable a rule', async () => {
    await eb.send(new DisableRuleCommand({ Name: ruleName }));
    expect((await eb.send(new DescribeRuleCommand({ Name: ruleName }))).State).toBe('DISABLED');

    await eb.send(new EnableRuleCommand({ Name: ruleName }));
    expect((await eb.send(new DescribeRuleCommand({ Name: ruleName }))).State).toBe('ENABLED');
  });

  // ──────────────────────────── Targets + PutEvents ────────────────────────────

  it('should create sink SQS queue', async () => {
    sinkQueueUrl = (await sqs.send(new CreateQueueCommand({ QueueName: uniqueName('eb-sink') }))).QueueUrl!;
    const attrs = await sqs.send(new GetQueueAttributesCommand({
      QueueUrl: sinkQueueUrl, AttributeNames: ['QueueArn'],
    }));
    sinkQueueArn = attrs.Attributes!['QueueArn'];
    expect(sinkQueueArn).toBeTruthy();
  });

  it('should put SQS target on rule', async () => {
    const response = await eb.send(new PutTargetsCommand({
      Rule: ruleName,
      Targets: [{ Id: 'sqs-target', Arn: sinkQueueArn }],
    }));
    expect(response.FailedEntryCount).toBe(0);
  });

  it('should list targets by rule', async () => {
    const response = await eb.send(new ListTargetsByRuleCommand({ Rule: ruleName }));
    const ids = response.Targets!.map(t => t.Id);
    expect(ids).toContain('sqs-target');
  });

  it('should deliver matching event to SQS target', async () => {
    await eb.send(new PutEventsCommand({
      Entries: [{
        Source: 'com.myapp',
        DetailType: 'OrderPlaced',
        Detail: JSON.stringify({ orderId: '123' }),
      }],
    }));

    const msg = await sqs.send(new ReceiveMessageCommand({
      QueueUrl: sinkQueueUrl,
      MaxNumberOfMessages: 1,
      WaitTimeSeconds: 2,
    }));
    expect(msg.Messages).toHaveLength(1);
    expect(msg.Messages![0].Body).toContain('com.myapp');
    expect(msg.Messages![0].Body).toContain('OrderPlaced');
  });

  it('should not deliver non-matching event', async () => {
    await eb.send(new PutEventsCommand({
      Entries: [{
        Source: 'other.app',
        DetailType: 'Ignored',
        Detail: '{}',
      }],
    }));

    const msg = await sqs.send(new ReceiveMessageCommand({
      QueueUrl: sinkQueueUrl,
      MaxNumberOfMessages: 1,
    }));
    expect(msg.Messages ?? []).toHaveLength(0);
  });

  // ──────────────────────────── InputTransformer ────────────────────────────

  it('should create transformer SQS queue', async () => {
    transformerQueueUrl = (await sqs.send(new CreateQueueCommand({ QueueName: uniqueName('eb-xform') }))).QueueUrl!;
    const attrs = await sqs.send(new GetQueueAttributesCommand({
      QueueUrl: transformerQueueUrl, AttributeNames: ['QueueArn'],
    }));
    transformerQueueArn = attrs.Attributes!['QueueArn'];
    expect(transformerQueueArn).toBeTruthy();
  });

  it('should put target with InputTransformer', async () => {
    const response = await eb.send(new PutTargetsCommand({
      Rule: ruleName,
      Targets: [{
        Id: 'transformer-target',
        Arn: transformerQueueArn,
        InputTransformer: {
          InputPathsMap: { src: '$.source', type: '$.detail-type' },
          InputTemplate: '{"source":"<src>","type":"<type>"}',
        },
      }],
    }));
    expect(response.FailedEntryCount).toBe(0);
  });

  it('should store InputTransformer correctly on target', async () => {
    const response = await eb.send(new ListTargetsByRuleCommand({ Rule: ruleName }));
    const xformTarget = response.Targets!.find(t => t.Id === 'transformer-target');
    expect(xformTarget).toBeDefined();
    expect(xformTarget!.InputTransformer).toBeDefined();
    expect(xformTarget!.InputTransformer!.InputPathsMap).toHaveProperty('src');
    expect(xformTarget!.InputTransformer!.InputTemplate).toContain('<src>');
  });

  it('should transform event payload before delivery', async () => {
    // Drain any prior messages
    await sqs.send(new ReceiveMessageCommand({
      QueueUrl: transformerQueueUrl, MaxNumberOfMessages: 10,
    }));

    await eb.send(new PutEventsCommand({
      Entries: [{
        Source: 'com.myapp',
        DetailType: 'OrderShipped',
        Detail: JSON.stringify({ orderId: '456' }),
      }],
    }));

    const msg = await sqs.send(new ReceiveMessageCommand({
      QueueUrl: transformerQueueUrl,
      MaxNumberOfMessages: 1,
      WaitTimeSeconds: 2,
    }));
    expect(msg.Messages).toHaveLength(1);
    const body = msg.Messages![0].Body!;
    expect(body).toContain('com.myapp');
    expect(body).toContain('OrderShipped');
    expect(body).not.toContain('orderId');
  });

  // ──────────────────────────── Tags ────────────────────────────

  it('should list tags for a rule resource', async () => {
    const response = await eb.send(new ListTagsForResourceCommand({ ResourceARN: ruleArn }));
    expect(response.Tags).toBeDefined();
  });

  // ──────────────────────────── Cleanup ────────────────────────────

  it('should remove targets', async () => {
    const response = await eb.send(new RemoveTargetsCommand({
      Rule: ruleName,
      Ids: ['sqs-target', 'transformer-target'],
    }));
    expect(response.FailedEntryCount).toBe(0);

    const list = await eb.send(new ListTargetsByRuleCommand({ Rule: ruleName }));
    expect(list.Targets ?? []).toHaveLength(0);
  });

  it('should delete the rule', async () => {
    await eb.send(new DeleteRuleCommand({ Name: ruleName }));
    await expect(eb.send(new DescribeRuleCommand({ Name: ruleName })))
      .rejects.toThrow();
  });

  it('should delete the event bus', async () => {
    await eb.send(new DeleteEventBusCommand({ Name: busName }));
    const list = await eb.send(new ListEventBusesCommand({}));
    expect(list.EventBuses!.map(b => b.Name)).not.toContain(busName);
  });
});
